src/app/util/attribute-table.cpp

```diff
@@ -100,8 +145,55 @@ static bool IsNullValue(const uint8_t * data, uint16_t dataLen, bool isAttribute
     return false;
 }
 
+namespace {
+/**
+ * Helper function to determine whether the attribute value for the given
+ * attribute is changing.  On success, the isChanging outparam will be set to
+ * whether the value is changing.
+ */
+Status AttributeValueIsChanging(EndpointId endpoint, ClusterId cluster, AttributeId attributeID, uint8_t * newValueData,
+                                EmberAfAttributeType dataType, bool * isChanging)
+{
+    size_t valueSize = AttributeTypeSize(dataType);
+
+    constexpr size_t maxValueSize = 16; // ipv6adr
+    if (valueSize > maxValueSize)
+    {
+        // Very much unexpected
+        ChipLogError(Zcl, "Attribute type %d has too-large size %u", dataType, static_cast<unsigned>(valueSize));
+        return Status::ConstraintError;
+    }
+
+    // valueSize will be 0 when we have no size information for dataType.
+    // In that case, we can't usefully read the current value, since we
+    // don't know how big it is.
+    if (valueSize == 0)
+    {
+        // To be safe, claim changing.
+        *isChanging = true;
+        return Status::Success;
+    }
+
+    uint8_t oldValueBuffer[maxValueSize];
+    // Cast to uint16_t is safe, because we checked valueSize <= maxValueSize above.
+    if (emberAfReadAttribute(endpoint, cluster, attributeID, oldValueBuffer, static_cast<uint16_t>(valueSize)) == Status::Success)
+    {
+        if (memcmp(newValueData, oldValueBuffer, valueSize) == 0)
```

Comment:
```cpp
*isChanging = (memcmp(newValueData, oldValueBuffer, valueSize) != 0);
```

may be clearer rather than relying on the fallback.  We seem to always be able to return. I would generally invert to make the exception/unhappy path "abort" and decrease indent

```cpp
if emberAfReadAttribute(endpoint, cluster, attributeID, oldValueBuffer, static_cast<uint16_t>(valueSize)) != Status::Success)
{
    *isChanging = true; // cannot read, so think we are changed
    return Status::Success;
}

*isChanging = (memcmp(newValueData, oldValueBuffer, valueSize) != 0);
return Status::Success;
```

Reply:
Done.